drv/stm32h7-sprot-server/src/main.rs

```diff
@@ -363,31 +364,88 @@ impl<S: SpiServer> Io<S> {
+            // Just unwrap this, because the `sys` task should never panic.
+            .unwrap_lite();
+
+        // Determine the deadline after which we'll give up, and start the clock.
+        let deadline = sys_get_timer()
+            .now
+            .checked_add(max_sleep as u64)
+            .unwrap_lite();
+        sys_set_timer(Some(deadline), notifications::sprot::TIMER_MASK);
+
         while self.is_rot_irq_asserted() != desired {
-            if slept == max_sleep {
+            // Wait to be notified either by the timeout or by the ROT_IRQ pin
+            // changing state.
+            const MASK: u32 =
+                notifications::sprot::TIMER_MASK | notifications::ROT_IRQ_MASK;
+            let notif = sys_recv_notification(MASK);
+
+            // First, check if the IRQ has fired. We do this by checking if the
+            // ROT_IRQ notification bit has been posted, and then asking the
+            // `sys` task to confirm that we actually got the IRQ.
+            //
+            // N.B. that we check this *before* checking for the timer
+            // notification bit, because it's possible *both* notifications were
+            // posted before we were scheduled again, and if the IRQ did fire,
+            // we'd prefer to honor that.
+            let irq_fired = notif & notifications::ROT_IRQ_MASK != 0
+                && self
+                    .sys
+                    .gpio_irq_control(
+                        notifications::ROT_IRQ_MASK,
+                        // If the IRQ hasn't fired, leave it enabled, otherwise,
+                        // if it has fired, don't re-enable the IRQ.
+                        IrqControl::Check,
+                    )
+                    // Sys task shouldn't panic.
+                    .unwrap_lite();
+            if irq_fired {
+                break;
+            }
+
+            // If the timer notification was posted, and the GPIO IRQ
+            // notification wasn't, we've waited for the timeout. Too bad!
+            if notif & notifications::TIMER_MASK != 0 {
+                // Disable the IRQ so that its notification doesn't go off
+                // again when we're not expecting it.
+                self.sys
+                    .gpio_irq_control(
+                        notifications::ROT_IRQ_MASK,
+                        IrqControl::Disable,
+                    )
+                    .unwrap_lite();
+
+                // Record the timeout.
                 self.stats.timeouts = self.stats.timeouts.wrapping_add(1);
                 ringbuf_entry!(Trace::RotReadyTimeout);
                 return false;
             }
-            hl::sleep_for(1);
-            slept += 1;
         }
+
+        // Ensure the timer gets unset before returning, to avoid frightening
+        // our IPC server when it's waiting in recv without expecting a timer to
+        // go off.
+        sys_set_timer(None, notifications::sprot::TIMER_MASK);
```

Comment:
Do we want to un-configure the interrupt? If this fn is called and the is_rot_irq_asserted() state is already the desired state then we drop to this point immediately.

Reply:
Ah, yeah, you're right, the IRQ being asserted when this function is called is the only case where we'll end up here with the IRQ still un-masked. Initially, I was performing that check before the IRQ was enabled and returning early, which is why I didn't mask it here. But, I moved it to the loop, and now it may still be enabled --- I'll add that.